### ISSUE TITLE
AWS storage upload issue

### ISSUE THREAD

**Issue by BalajiSriraman** [aws-amplify/amplify-js#13409]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

Storage

### Amplify Version

v6

### Amplify Categories

storage

### Backend

None

### Environment information

<details>

```
# Put output below this line
 OS: Linux 6.8 Pop!_OS 22.04 LTS
    CPU: (16) x64 11th Gen Intel(R) Core(TM) i9-11900K @ 3.50GHz
    Memory: 42.07 GB / 62.42 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 18.18.0 - ~/.nvm/versions/node/v18.18.0/bin/node
    npm: 9.8.1 - ~/.nvm/versions/node/v18.18.0/bin/npm
    pnpm: 9.1.0 - ~/.nvm/versions/node/v18.18.0/bin/pnpm
  Browsers:
    Chrome: 125.0.6422.76
  npmPackages:
    @aws-amplify/backend: ^1.0.2 => 1.0.2 
    @aws-amplify/ui-react: ^6.1.9 => 6.1.9 
    @aws-amplify/ui-react-core: ^3.0.14 => 3.0.14 
    @heroicons/react: ^2.1.3 => 2.1.3 
    @types/react: ^18.2.66 => 18.3.2 
    @types/react-dom: ^18.2.22 => 18.3.0 
    @typescript-eslint/eslint-plugin: ^7.2.0 => 7.8.0 
    @typescript-eslint/parser: ^7.2.0 => 7.8.0 
    @vitejs/plugin-react-swc: ^3.5.0 => 3.6.0 
    autoprefixer: ^10.4.19 => 10.4.19 
    aws-amplify: ^6.3.0 => 6.3.0 
    clsx: ^2.1.1 => 2.1.1 
    eslint: ^8.57.0 => 8.57.0 
    eslint-plugin-react-hooks: ^4.6.0 => 4.6.2 
    eslint-plugin-react-refresh: ^0.4.6 => 0.4.7 
    framer-motion: ^10.10.0 => 10.18.0 
    postcss: ^8.4.38 => 8.4.38 
    react: ^18.2.0 => 18.3.1 
    react-dom: ^18.2.0 => 18.3.1 
    react-dropzone: ^14.2.3 => 14.2.3 
    react-icons: ^4.8.0 => 4.12.0 
    react-router-dom: ^6.9.0 => 6.23.1 
    react-toastify: ^9.1.2 => 9.1.3 
    tailwind-merge: ^2.3.0 => 2.3.0 
    tailwindcss: ^3.4.3 => 3.4.3 
    typescript: ^5.2.2 => 5.4.5 
    vite: ^5.2.0 => 5.2.11 
  npmGlobalPackages:
    @withgraphite/graphite-cli: 1.3.4
    corepack: 0.19.0
    nodemon: 3.1.0
    npm: 9.8.1
    pnpm: 9.1.1
    typescript: 5.4.5

```

</details>

### Describe the bug

trying to upload a APK file to S3 bucket but facing Multipart issue
![image](https://github.com/aws-amplify/amplify-js/assets/96479581/b1b4c871-89d6-499b-af14-c5dff180226c)

### Expected behavior

Just normal file upload

### Reproduction steps
**Code**
```js

>  const handleFileSelect = async (file: File) => {
>     try {
>       const awsfile = uploadData({
>         data: file,
>         path({ identityId }) {
>           return `uploads/${identityId}/test_${file.name}`;
>         },
>         options: {
>           // metadata: {
>           //   date: `${Date.now()}`,
>           // },
>           contentEncoding: "base64",
>           contentType: "application/vnd.android.package-archive",
>           contentDisposition: "attachment",
>           onProgress: ({ transferredBytes, totalBytes }) => {
>             if (totalBytes) {
>               console.log(
>                 `Upload progress ${Math.round(
>                   (transferredBytes / totalBytes) * 100
>                 )} %`
>               );
>             }
>           },
>         },
>       }).result;
> 
>       console.log(awsfile.then((res) => console.log(res)));
>     } catch (err) {
>       console.log(err);
>     }
> 
```

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

**Comment by BalajiSriraman** [comment]
Hi @cwomack  

 it usually small around 9mb .apk file, but issue persists with files of size 300mb also

upload is quite quick, have got a 30mbps connection so around 3 sec for 9mb file

**Comment by cwomack** [comment]
@BalajiSriraman, thank you for the additional context.  We'll try to reproduce on our side and see if we can root cause the issue.

**Comment by jl4nz** [comment]
I have the same issue, randomly. For small files on KB doesn't seem to happen, but when using this library on MB files (pdf 6MB in my case) ,  the `uploader.result` promise gets randomly rejected and some times the same file returns `status:"fulfilled"`.

I'm using a bucket created externally from Amplify CLI... but added the exposedHeaders: `'x-amz-server-side-encryption', 'x-amz-request-id', 'x-amz-id-2', 'ETag'`

<img width="1489" alt="Screenshot 2024-07-30 at 16 26 10" src="https://github.com/user-attachments/assets/1546672c-8f42-44e4-9a33-4fdd00c417c7">

**Comment by jl4nz** [comment]
After some more investigation, I got this working after the bucket is configured with the following CORS config.

```json
[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "POST",
            "GET",
            "HEAD",
            "DELETE",
            "PUT"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": [
            "ETag",
            "x-amz-server-side-encryption",
            "x-amz-request-id",
            "x-amz-id-2"
        ],
        "MaxAgeSeconds": 3000
    }
]
```

Maybe this can be an issue of just adding to the docs that when importing a S3 bucket, add the correct config.

**Comment by antoineaws** [comment]
Same issue here, multi-part upload fails with version 6.5.3 with the above error `Error: Invalid parameter for ComplteMultipartUpload API: [object Object]` (note typo as well error message - CompleteMultipartUpload. In my case user only has s3:PutObject permissions on bucket. Will try the above settings on the bucket and see if it makes a difference.

**Comment by TheForsakenSpirit** [comment]
@cwomack Hi! I have exactly same issue. Start on 5mb files(when amplify switches to multipart upload). Do you need some aditional data to reproduce and fix it? Also throwing error with `[object Object]` message not a good practice to be fair.

**Comment by AllanZhengYP** [comment]
hi @TheForsakenSpirit 

Thank you for reporting this. Can you confirm if you are using the latest version of Amplify JS library? The error message seems to be from the S3 server side. Can you share the redacted version of the **request payload** of the errored request.

**Comment by TheForsakenSpirit** [comment]
@AllanZhengYP yep. We use latest amplify. All requests before resolved with 200.
I traced error to missing `ETag` field on parts upload when sdk build upload completed request.
All requests before have ETag in headers.
Will add screenshots tomorrow.

**Comment by TheForsakenSpirit** [comment]
![Image](https://github.com/user-attachments/assets/91a41749-55fb-449f-997f-4352195d6eac)

![Image](https://github.com/user-attachments/assets/38c4ef2d-8883-4274-af01-d798994186bd)

![Image](https://github.com/user-attachments/assets/5d9ef740-a6b2-4d68-b1c8-385613e129c8)

Errors throws [here](https://github.com/aws-amplify/amplify-js/blob/8b489d1667c67741cbc1111b0a52d10996aa857d/packages/storage/src/providers/s3/utils/client/s3data/completeMultipartUpload.ts#L106-L109), becouse ETag missing, but it is exist in response headers.

For some reason uploadPart function recive next response:
```
{
    "statusCode": 200,
    "headers": {
        "content-length": "0",
        "": ""
    },
    "body": {},
    "$metadata": {
        "attempts": 1
    }
}
```

**Comment by cwomack** [comment]
@TheForsakenSpirit, thank you for the screenshots. Can you also share any minimal code sample for the `uploadData()` call while we reproduce this on our side as well?

**Comment by AllanZhengYP** [comment]
@TheForsakenSpirit 

Can you confirm you have followed this this guidance to add the CORS configuration to your bucket? Especially around the `"ExposeHeaders"`. You need to make sure `Etag` header is exposed: https://docs.amplify.aws/react/build-a-backend/storage/extend-s3-resources/

**Comment by TheForsakenSpirit** [comment]
@AllanZhengYP @cwomack yeap. That the headers in CORS (config has empty arrays) . THANKS!!! Now it works.
Can you add more detailed info in error?

**Comment by cshfang** [comment]
@TheForsakenSpirit Good to hear that your issue was resolved!

Given the number of people running into this, we have updated the error messaging (https://github.com/aws-amplify/amplify-js/pull/14170) to hopefully mitigate future developers having to spin their wheels too much over this same problem without any hints.
